README.md
# WebMo
[![Bundle size](https://img.shields.io/bundlephobia/minzip/webmo.svg)](https://bundlephobia.com/result?p=webmo)
[![npm version](https://badge.fury.io/js/webmo.svg)](https://www.npmjs.com/package/webmo)
[![TypeScript](https://img.shields.io/badge/%3C%2F%3E-TypeScript-blue.svg)](http://www.typescriptlang.org/)
[![License](https://img.shields.io/npm/l/webmo.svg)](https://github.com/evanshortiss/webmo/blob/master/package.json)

A tiny library to capture device motion, rotation, and orientation events.

## iOS 12.2+ Security
From iOS 12.2 onwards the Device Motion and Orientation APIs are disabled by
default and can only be used by HTTPS domains. You need to enable access to
the APIs in `Settings > Safari > Privacy & Security`.

This might change in a future release to instead use a popup similar to the
Geolocation API permission access.


## Usage

```ts
import * as webmo from 'webmo'

async function startCapture () {
  const hasMotionSupport = await webmo.motion.deviceHasMotionSupport()
  
  if (hasMotionSupport) {
    const mListener = new webmo.motion.MotionListener(data => {
      console.log('Received motion payload: ', data)
    })
  }
  
  const hasOrientationSupport = await webmo.orientation.deviceHasOrientationSupport()
  
  if (hasOrientationSupport) {
    const oListener = new webmo.orientation.OrientationListener(data => {
      console.log('Received orientation payload: ', data)
    })
  }
}

startCapture()
```

## Behaviours

Behaviour like this can be made configurable through options in the future.

### Alpha Orientation Value
This library normalises device alpha orientation, i.e the position the phone is in when the first
orientation event is captured is used as alpha of 0. This is only necessary on Android devices, iOS devices default to 0 as the initial alpha heading.

### Device Support Tests
The `orientation.deviceHasOrientationSupport` and `motion.deviceHasMotionSupport` functions are designed to detect a motion event. If none is detected then you might get a false negative, e.g user left their phone lying on a table so no motion change is detected. Make sure you account for this in your application, by telling the user to "hold the device in their hand" or similar.


## Example

To try the example application do the following:

```
git clone https://github.com/evanshortiss/webmo.git
cd webmo
npm i
npm start
```

Now visit port 8080 using a device that supports motion and orientation, for example a modern iOS or Android phone, and you'll see live data from the sensors.

## API

### webmo.motion

All motion related functions fall under this namespace.

### webmo.motion.mayHaveMotionSupport(): boolean

Returns `true` if `DeviceMotionEvent` is defined globally.

### webmo.motion.deviceHasMotionSupport(timeout: number): Promise\<boolean>

Tests for motion support by detecting motion events. Defaults to a 1000 millisecond timeout before resolving with `false`. Resolves with `true` if a motion event is detected before `timeout` is reached.

### webmo.motion.MotionListener(callback, options)

A class that can be instantiated to listen for motion events. Accepts two arguments:

* callback - Callback that receives sensor data
* options - Object containing configuration

Valid options are:

* threshold (Default: 0) - Threshold required to register motion. Can be used to filter out smaller motions.
* rotationRateThreshold (Default: 0) - Threshold required to fire an event for rotation rate. Can be used to filter out smaller/sloewr rotations.
* autoStart (Default: true) - If motion events should be listened for immediately.

Callback and data format:

```ts
const options = {
  threshold: 5
}
const listener = new webmo.MotionListener((data) => {
  // data is an object like so:
  // {
  //   acceleration: {
  //     x: number
  //     y: number
  //     z: number
  //   }
  //   rotationRate: {
  //     alpha: number
  //     beta: number
  //     gamma: number
  //   }
  //   timestamp: number
  // }
}, options)
```

#### webmo.motion.MotionListener.start()

Start listening for motion events. Only required if `options.autoStart` was set to false.

#### webmo.motion.MotionListener.stop()

Stop listening for motion events.

#### webmo.motion.MotionListener.setListener(callback: function|undefined) 

Replace the listener function that was passed to the instance constructor.

#### webmo.motion.MotionListener.setThreshold(n: number)

Changes the threshold required for this instance to consider a motion or rotation event event worth emitting.

#### webmo.motion.MotionListener.isListening(): boolean

Returns `true` if the instance has bound event handlers for motion.

### webmo.orientation

The orientation API is the same as `webmo.motion`, but returned data objects have the following structure:

```ts
const listener = new webmo.MotionListener((data) => {
  // data is an object like so:
  // {
  //   alpha: number
  //   beta: number
  //   gamma: number
  //   timestamp: number
  // }
})
```


CHANGELOG.md
# CHANGELOG

## 03/04/19 - 0.5.2
* Fix return type annotion for device support tests

## 02/04/19 - 0.5.1
* Add iOS specific notes about security settings

## 20/03/19 - 0.5.0
* Improve documentation
* Add helpful badges in README

## 14/03/19 - 0.4.0

### Public API Changes
* Add `rotationRate` data to motion events
* Add `rotationRateThreshold` option for motion constructor

### Internal Changes
* Standardise event verification in `AbstractListener` and subclasses

## 05/03/19 - 0.3.0

* Add `setListener` function instances

## 19/02/19 - 0.2.0

### Public API Changes
* Add `getLastCapturedEvent` method to listeners
* Make `listener` argument optional for constructors
* Export `ListenerOptions` interface

### Internal Changes
* Add `formatEvent()` method to `AbstractListener`
* Add `fireEvent()` method to `AbstractListener`
* Implement `onChangeEvent()` method in `AbstractListener`


src/motion.ts
import { AbstractListener } from './abstract-listener';

const EVT_NAME = 'devicemotion';

/**
 * Determies if DeviceMotionEvent is available for use.
 * This does not indicate that the device actually supports motion.
 */
export function mayHaveMotionSupport() {
  return typeof DeviceMotionEvent !== 'undefined';
}

/**
 * Determines if motion support is available by listening for a motion event.
 * Waits up to 500 milliseconds for an event by default, but can be extended.
 * @param timeout Time to wait before deciding motion support is unavailable.
 */
export function deviceHasMotionSupport(timeout = 1000): Promise<boolean> {
  return new Promise(resolve => {
    const listener = (e: DeviceMotionEvent) => {
      window.removeEventListener(EVT_NAME, listener);
      if (e.acceleration) {
        resolve(true);
      } else {
        resolve(false);
      }
    };

    window.addEventListener(EVT_NAME, listener);

    setTimeout(() => resolve(false), timeout);
  });
}

export interface MotionListenerEvent {
  rotationRate: {
    alpha: number;
    beta: number;
    gamma: number;
  };
  acceleration: {
    x: number;
    y: number;
    z: number;
  };
  timestamp: number;
}

export interface VerifiedMotionEvent extends DeviceMotionEvent {
  acceleration: {
    x: number;
    y: number;
    z: number;
  };
  rotationRate: {
    alpha: number;
    beta: number;
    gamma: number;
  };
}

export class MotionListener extends AbstractListener<
  MotionListenerEvent,
  DeviceMotionEvent,
  VerifiedMotionEvent
> {
  protected eventName = EVT_NAME;

  verifyEventStructure(e: DeviceMotionEvent) {
    if (
      e.acceleration &&
      e.acceleration.x &&
      e.acceleration.y &&
      e.acceleration.z &&
      e.rotationRate &&
      e.rotationRate.alpha &&
      e.rotationRate.beta &&
      e.rotationRate.gamma
    ) {
      return e as VerifiedMotionEvent;
    } else {
      throw new Error(
        'The "acceleration" or "rotationRate" property in DeviceMotionEvent appears to be malformed. This library only works on devices that fully support this event.'
      );
    }
  }

  private isAboveThreshold(e: VerifiedMotionEvent) {
    if (this.options.threshold) {
      const { x, y, z } = e.acceleration;

      return (
        Math.abs(x) > this.options.threshold ||
        Math.abs(y) > this.options.threshold ||
        Math.abs(z) > this.options.threshold
      );
    } else {
      return true;
    }
  }

  private isAboveRotationRateThreshold(e: VerifiedMotionEvent) {
    if (this.options.rotationRateThreshold) {
      const { alpha, beta, gamma } = e.rotationRate;

      return (
        Math.abs(alpha) > this.options.rotationRateThreshold ||
        Math.abs(beta) > this.options.rotationRateThreshold ||
        Math.abs(gamma) > this.options.rotationRateThreshold
      );
    } else {
      return true;
    }
  }

  isChangeAboveThreshold(e: VerifiedMotionEvent) {
    return this.isAboveThreshold(e) || this.isAboveRotationRateThreshold(e);
  }

  formatEvent(e: VerifiedMotionEvent) {
    return {
      rotationRate: e.rotationRate,
      acceleration: e.acceleration,
      timestamp: Date.now()
    };
  }
}


src/abstract-listener.ts
export interface ListenerOptions {
  threshold?: number;
  rotationRateThreshold?: number;
  autoStart?: boolean;
}

export abstract class AbstractListener<
  WrappedDomEvent,
  BaseEvent,
  VerifiedEvent extends BaseEvent
> {
  protected eventName = '';
  protected _previousEvent: WrappedDomEvent | undefined;
  private _internalListener = (e: any) => this.onChangeEvent(e);
  private _listening = false;

  constructor(
    protected listener?: (data: WrappedDomEvent) => void,
    protected options: ListenerOptions = {}
  ) {
    this.options = Object.assign({ autoStart: true }, options);

    if (this.options.autoStart) {
      setTimeout(() => this.start());
    }
  }

  /**
   * Start listening for motion events.
   */
  public start() {
    if (!this._listening) {
      // Prevent adding same listener multiple times...
      window.addEventListener(this.eventName, this._internalListener);
      this._listening = true;
    }
  }

  /**
   * Stop listening for motion events.
   */
  public stop() {
    window.removeEventListener(this.eventName, this._internalListener);
    this._listening = false;
  }

  /**
   * Change the threshold required to record a motion event.
   * Any motion that registers below this value on all axes is ignored.
   * @param threshold
   */
  public setThreshold(threshold: number) {
    this.options.threshold = threshold;
  }

  /**
   * Returns the latest captured event if one is available.
   */
  public getLastCapturedEvent() {
    return this._previousEvent;
  }

  /**
   * Returns the status of this listener.
   */
  public isListening() {
    return this._listening;
  }

  /**
   * Set the listener callback function on this instance.
   * Can pass undefined to remove a listener.
   * @param listener
   */
  public setListener(listener?: (data: WrappedDomEvent) => void) {
    this.listener = listener;
  }

  /**
   * Determines if the reported event is great enough to warrant
   * attention based on the defined threshold.
   * @param e
   */
  protected abstract isChangeAboveThreshold(data: VerifiedEvent): boolean;

  /**
   * Formats a captured event to comply with the custom structure defined
   * by the listener.
   * @param e
   */
  protected abstract formatEvent(e: VerifiedEvent): WrappedDomEvent;

  /**
   * Accepts a browser event, performs verification, and returns an
   * event that is verified. This is to avoid excessive null checks.
   * @param e
   */
  protected abstract verifyEventStructure(e: BaseEvent): VerifiedEvent;

  /**
   * Fires the listener, if one was provided to the constructor.
   * @param e
   */
  protected fireListener(e: WrappedDomEvent) {
    if (this.listener) {
      this.listener(e);
    }
  }

  /**
   * Internal function used to determine if the orientation callback will be
   * fired.
   * @param e
   */
  protected onChangeEvent(e: BaseEvent) {
    const verifiedEvent = this.verifyEventStructure(e);

    if (this.isChangeAboveThreshold(verifiedEvent)) {
      this._previousEvent = this.formatEvent(verifiedEvent);
      this.fireListener(this._previousEvent);
    }
  }
}


src/orientation.ts
import { AbstractListener } from './abstract-listener';

const EVT_NAME = 'deviceorientation';

/**
 * Determies if DeviceOrientationEvent is available for use.
 * This does not indicate that the device actually supports motion.
 */
export function mayHaveOrientationSupport() {
  return typeof DeviceOrientationEvent !== 'undefined';
}

/**
 * Determines if motion support is available by listening for a motion event.
 * Waits up to 500 milliseconds for an event by default, but can be extended.
 * @param timeout Time to wait before deciding motion support is unavailable.
 */
export function deviceHasOrientationSupport(timeout = 1000): Promise<boolean> {
  return new Promise(resolve => {
    const listener = (e: DeviceOrientationEventInit) => {
      window.removeEventListener(EVT_NAME, listener);
      resolve(true);
    };

    window.addEventListener(EVT_NAME, listener);

    setTimeout(() => resolve(false), timeout);
  });
}

export interface OrientationListenerEvent {
  alpha: number;
  beta: number;
  gamma: number;
  absolute?: boolean;
  timestamp: number;
}

export interface VerifiedOrientationEvent extends DeviceOrientationEvent {
  alpha: number;
  beta: number;
  gamma: number;
}

export class OrientationListener extends AbstractListener<
  OrientationListenerEvent,
  DeviceOrientationEvent,
  VerifiedOrientationEvent
> {
  isAndroidDevice = navigator.userAgent.match(/android/gi) !== null;
  initialEvent: DeviceOrientationEvent | undefined = undefined;
  eventName = EVT_NAME;

  private getAlpha(alpha: number): number {
    if (this.isAndroidDevice && this.initialEvent && this.initialEvent.alpha) {
      let a = alpha - this.initialEvent.alpha;

      if (a < 0) {
        a += 360;
      }

      return a;
    } else {
      return alpha;
    }
  }

  verifyEventStructure(e: DeviceOrientationEvent) {
    if (
      typeof e.alpha !== 'number' ||
      typeof e.beta !== 'number' ||
      typeof e.gamma !== 'number'
    ) {
      throw new Error(
        'The "alpha", "beta", or "gamma" property from DeviceOrientationEvent appears to be malformed. This library only works on devices that fully support this event.'
      );
    } else {
      return e as VerifiedOrientationEvent;
    }
  }

  isChangeAboveThreshold(data: VerifiedOrientationEvent): boolean {
    const previous = this._previousEvent;

    if (!previous) {
      return true;
    } else if (this.options.threshold) {
      const { alpha, beta, gamma } = data;
      return (
        Math.abs(alpha - previous.alpha) > this.options.threshold ||
        Math.abs(beta - previous.beta) > this.options.threshold ||
        Math.abs(gamma - previous.gamma) > this.options.threshold
      );
    } else {
      return true;
    }
  }

  formatEvent(e: VerifiedOrientationEvent) {
    if (!this._previousEvent) {
      // Kind of nasty hack.
      // We're saying the current evt is also initial. Whatever...
      this.initialEvent = e;
    }

    return {
      alpha: this.getAlpha(e.alpha),
      beta: e.beta,
      gamma: e.gamma,
      timestamp: Date.now()
    };
  }
}
